### PR TITLE
Further Expand .simulation() capabilities in LGBN's

### LINKED ISSUE #2110: [ENH] Expand simulation capability of Linear Gaussian Bayesian Networks
The `LinearGaussianBayesianNetwork.simulate` method currently provides some basic functionality to generate data from the model. Compared to `DiscreteBayesianNetwork`, this function is quite limiting as it doesn't allow users to specify any additional arguments such as evidence, intervention, or missing data. We would like to expand this simulation method to have all the functionality that the DiscreteBN.simulate offers.

- [x] Specifying evidence: Use the predict method to compute the posterior distribution and sample from the posterior distribution.
- [x] Specifying interventions: Modify the network structure and the LinearGaussianCPDs according to the specified intervention and simulate data from the modified model.
- [ ] Specifying virtual interventions: Similar to interventions, but assign the intervened node to the specified CPD.
- [ ] Add `include_latents` argument. If True, returns the full simulated dataframe otherwise drops the latent variables.
- [ ] Missing data: Requires modification to the model by adding new nodes that control the missing values.  See how it is implemented in DiscreteBN.

### PR DESCRIPTION
**Pull request by Sanchay117**
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2110 

### List of changes to the codebase in this pull request
- Added support for virtual interventions
- Added a test as well for virtual_interventions

### CODE REVIEW

**Comment by Sanchay117** [comment]
Hey @ankurankan, could you please confirm if my approach for handling `virtual_interventions` is correct?
Currently, I’m replacing the CPDs of the specified variables with those provided in the `virtual_interventions` list.
I’m not entirely sure if this is the right way to handle it. If it looks good to you, I’ll go ahead and start implementing the remaining features.

**Comment by Sanchay117** [comment]
Also @ankurankan, there seems to be an inconsistency between the implementation in `LinearGaussianCPD.py` and the tests in `test_LinearGaussianBayesianNetwork.py`:

In the `setUp` method of the tests, we define the CPDs like this:

```python
def setUp(self):
    self.model = LinearGaussianBayesianNetwork([("x1", "x2"), ("x2", "x3")])
    self.cpd1 = LinearGaussianCPD(variable="x1", beta=[1], std=4)
    self.cpd2 = LinearGaussianCPD(
        variable="x2", beta=[-5, 0.5], std=4, evidence=["x1"]
    )
    self.cpd3 = LinearGaussianCPD(
        variable="x3", beta=[4, -1], std=3, evidence=["x2"]
    )
```

Here, we specify the standard deviations directly.
However, in the `test_simulate` method:
```python
def test_simulate(self):
    self.model.add_cpds(self.cpd1, self.cpd2, self.cpd3)
    df_cont = self.model.simulate(n_samples=10000, seed=42)

    # Same model in terms of equations
    rng = np.random.default_rng(seed=42)
    x1 = 1 + rng.normal(0, 2, 10000)
    x2 = -5 + 0.5 * x1 + rng.normal(0, 2, 10000)
    x3 = 4 + -1 * x2 + rng.normal(0, np.sqrt(3), 10000)
    df_equ = pd.DataFrame({"x1": x1, "x2": x2, "x3": x3})

    np_test.assert_array_almost_equal(df_cont.mean(), df_equ.mean(), decimal=1)
    np_test.assert_array_almost_equal(df_cont.cov(), df_equ.cov(), decimal=1)
```
We are passing values like `2` and `np.sqrt(3)` as the standard deviations to `rng.normal`, which expects the standard deviation as input.

This suggests that the `std` parameter passed to `LinearGaussianCPD` might actually be interpreted as variance in the implementation, even though it's named `std`. It would be great if you could clarify whether this is intentional or a mismatch between the code and the tests.

I have opened an issue regarding this as well.

**Comment by Sanchay117** [comment]
And, it appears the tests failed because I forgot to run the linting checks. My apologies!

**Comment by Sanchay117** [comment]
Also, I feel that this error check could be more robust:
```python
if len(self.cpds) != len(self.nodes()):
            raise ValueError(
                "Each node in the model should have a CPD associated with it"
            )
```
Instead of checking the lengths of the CPD list and the number of nodes which might still pass even if the CPDs correspond to the wrong variables we could instead verify that each node in the model has exactly one CPD associated with it.

Please let me know if you’d like me to proceed with this change.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 1]
```diff
@@ -407,6 +422,23 @@ def simulate(
         else:
             model = self
 
+        # Step 3: If virtual_interventions are specified, change the CPD's of intervened variables
+        if virtual_intervention != []:
+            # Step 3.1: Check if CPD's are valid
+            for cpd in virtual_intervention:
+                var = cpd.variable
+                if var not in model.nodes():
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Please move all checks to the beginning of the method in Step 1.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 2]
```diff
@@ -407,6 +422,23 @@ def simulate(
...
+                    raise ValueError(
+                        f"Virtual intervention provided for variable which is not in the model: {var}"
+                        f"The following nodes are present in the model: {model.nodes()}"
+                    )
+
+            for cpd in virtual_intervention:
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: We would also need to remove all incoming edges to the variables. This currently is not a valid model as every CPD in the model needs to be `P(var | parents(var))`, however, here `var` has parents but the CPD is just `P(var)`

**Review by ankurankan — CHANGES_REQUESTED** [review]
I have added a comment about removing all incoming edges to the virtual intervention variable. This is similar to the model structure modification that we do for `do` variables.

**Comment by ankurankan** [comment]
> Also, I feel that this error check could be more robust:
> 
> ```python
> if len(self.cpds) != len(self.nodes()):
>             raise ValueError(
>                 "Each node in the model should have a CPD associated with it"
>             )
> ```
> 
> Instead of checking the lengths of the CPD list and the number of nodes which might still pass even if the CPDs correspond to the wrong variables we could instead verify that each node in the model has exactly one CPD associated with it.
> 
> Please let me know if you’d like me to proceed with this change.

Thanks for pointing this out. Please feel free to fix this.

**Comment by Sanchay117** [comment]
> > Also, I feel that this error check could be more robust:
> > ```python
> > if len(self.cpds) != len(self.nodes()):
> >             raise ValueError(
> >                 "Each node in the model should have a CPD associated with it"
> >             )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Instead of checking the lengths of the CPD list and the number of nodes which might still pass even if the CPDs correspond to the wrong variables we could instead verify that each node in the model has exactly one CPD associated with it.
> > Please let me know if you’d like me to proceed with this change.
> 
> Thanks for pointing this out. Please feel free to fix this.

It appears there was a method `check_model` for the same purpose but it wasn't being used so I am using it now.

File: pgmpy/models/LinearGaussianBayesianNetwork.py (hunk 1 above)
Comment (Sanchay117) [reply pgmpy/models/LinearGaussianBayesianNetwork.py]: Done.

File: pgmpy/models/LinearGaussianBayesianNetwork.py (hunk 2 above)
Comment (Sanchay117) [reply pgmpy/models/LinearGaussianBayesianNetwork.py]: So now, I am assigning the new CPD's to their respctive variables and removing the incoming edges as well. Please let me know if this is the correct approach?

Comment (ankurankan) [reply pgmpy/models/LinearGaussianBayesianNetwork.py]: Indeed. In the case of hard intervention, we had to remove the node and change the CPD for the child nodes because there is no way to define a Gaussian distribution that has the whole probability mass at a single value.

Comment (Sanchay117) [reply pgmpy/models/LinearGaussianBayesianNetwork.py]: Okay, thank you for the explanation. I have done this only in the latest commit.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 3]
```diff
@@ -409,15 +432,21 @@ def simulate(
 
                 model.remove_node(var)
 
-        else:
-            model = self
+        # Step 3: If virtual_interventions are specified, change the CPD's of intervened variables
+        # to specified ones and remove the incoming nodes
+        if virtual_intervention != []:
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Don't really need this check. Right? If it is an empty list, the code block would just get skipped.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 4]
```diff
@@ -453,6 +482,10 @@ def simulate(
         for do_var, do_val in do.items():
             df[do_var] = do_val
 
+        # Step 6: Remove latent variables if specified
+        if not include_latents:
+            df = df.drop(columns=self.latents, errors="ignore")
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Why would we need `errors='ignore'`? I don't think we will ever run into a situation where any of the latents is not present in the dataframe.

**Review by ankurankan — CHANGES_REQUESTED** [review]
I think it looks good. I have added a couple of minor comments.

File: pgmpy/models/LinearGaussianBayesianNetwork.py (hunk 4 above)
Comment (Sanchay117) [reply pgmpy/models/LinearGaussianBayesianNetwork.py]: You are right, I have removed that now.

File: pgmpy/models/LinearGaussianBayesianNetwork.py (hunk 3 above)
Comment (Sanchay117) [reply pgmpy/models/LinearGaussianBayesianNetwork.py]: I have removed this check now.

**Comment by Sanchay117** [comment]
Hi @ankurankan,

Here is what I have done to handle `missing_prob`:

### For each CPD in `missing_prob`:

- Add the missingness variable (the CPD’s variable, e.g. `X*`) as a **new node** to the model.
- Add edges **from each evidence node to this missingness node**.

### Sampling the data:

- Perform the normal simulation/sampling step to generate data for **all variables, including the missingness variables**.
- This produces a DataFrame containing columns for the original variables (e.g., `X`) **and** their corresponding missingness indicators (e.g., `X*`).

### Applying the missingness mask:

For each missingness variable `var*`:

1. Use the simulated values of `var*` as **logits** (log-odds) for the probability of missingness.
2. Convert logits to probabilities using the sigmoid function:

   p_missing = 1 / (1 + exp(-logit))

3. For each sample, draw from a Bernoulli distribution with probability p_missing to decide if that sample of the original variable `var` should be missing.
4. Create a new column `var_full` that stores the original unmasked values (before applying missingness).
5. Set values of `var` to `NaN` wherever the Bernoulli draw indicates missingness.
6. Drop the missingness indicator column `var*` from the final DataFrame.

---

Please let me know if this is the correct approach.

**Comment by Sanchay117** [comment]
Also, if there’s any other issue you think I could help with, I’d really appreciate it if you could point me to one. I’d love to pick something up, as most open issues seem to be already assigned or have PRs.

**Comment by ankurankan** [comment]
@Sanchay117 can you remove the missingness part from this PR? I am not yet sure how missingness can be handled in the GaussianBN case and will need some time to think about it. The remaining part of the PR looked good, so if you remove the missingness part for now, we can merge the rest.

**Comment by Sanchay117** [comment]
@ankurankan Sure, I'll remove the `missing_prob` part.

**Comment by ankurankan** [comment]
@Sanchay117 I think another missing part in simulations is `virtual_evidence`. This would be similar to the `evidence` argument, where we compute a conditional distribution using the evidence values and sample from this conditional distribution. The difference would be that the evidence variable would need to be set to a distribution instead of a single value. I think there should be straightforward formulas to compute that, but I haven't looked it up. Another thing to consider here would be handling both hard and virtual evidence together. Would you like to take this up? If yes, feel free to create another PR for that, and check if this is already in the list of features in the issue.

**Comment by Sanchay117** [comment]
@ankurankan I'd love to take this up! I'll create a separate PR for it. Just checked #2110, and `virtual_evidence` doesn’t seem to be listed as a feature there.

**Comment by Sanchay117** [comment]
@ankurankan , I have removed the missingness part. You can merge the PR now if everything looks right to you. I'll make another request for `virtual_evidence`.

**Review by ankurankan — APPROVED** [review]
Thanks for the PR @Sanchay117
